From cfa76d7fd3b3a6e974c57a325bf70e37227b0222 Mon Sep 17 00:00:00 2001
From: Doug Smith <Doug.Smith@lairdtech.com>
Date: Fri, 11 Jul 2014 15:21:16 -0400
Subject: [PATCH] BZ5784: fips bridging mode failing

Bridging mode transmit packet failed
due to insufficient headroom.

Modified to copy and expand the transmit skb
if cloned, or insufficient head/tailroom
according to dev->needed_head/tailroom.

Bug: 5784
---
 drivers/net/wireless/laird_fips/laird.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/laird_fips/laird.c b/drivers/net/wireless/laird_fips/laird.c
index 564036a..1e6c5ec 100644
--- a/drivers/net/wireless/laird_fips/laird.c
+++ b/drivers/net/wireless/laird_fips/laird.c
@@ -358,10 +358,12 @@ int laird_skb_tx_prep(struct sk_buff *skbin, struct net_device *dev, int wmm,
 		return -1;
 	}
 
-	/* TBD: move this into the other routine... */
-	if (skb_cloned(skb) || skb_tailroom(skb) < 8) {
-		/* tailroom is too small -- try making a copy with more tailroom */
-		skb = skb_copy_expand(skb, skb_headroom(skb), 8, GFP_ATOMIC);
+	if (skb_cloned(skb) ||
+		skb_headroom(skb) < dev->needed_headroom ||
+		skb_tailroom(skb) < dev->needed_tailroom)
+	{
+		/* make a copy -- cloned, or insufficient head/tail room */
+		skb = skb_copy_expand(skb, dev->needed_headroom, dev->needed_tailroom, GFP_ATOMIC);
 		if (!skb) {
 			skb = skbin;
 			res = -ENOMEM;
-- 
1.8.3.2

